core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -302,6 +308,46 @@ public static RexNode splitFilter(final RexSimplify simplify,
     return null;
   }
 
+  /**
+   * Reorders some of the operands in this expression so structural comparison,
+   * i.e., based on string representation, can be more precise.
+   */
+  private static RexNode rewriteNode(RexBuilder rexBuilder, RexNode condition) {
+    switch (condition.getKind()) {
+    case AND:
+    case OR: {
+      RexCall call = (RexCall) condition;
+      List<Pair<String, RexNode>> newOperands = new ArrayList<>();
+      for (RexNode operand : call.operands) {
+        operand = rewriteNode(rexBuilder, operand);
+        newOperands.add(Pair.of(operand.toString(), operand));
+      }
+      Collections.sort(newOperands, new Comparator<Pair<String, RexNode>>() {
+        @Override public int compare(Pair<String, RexNode> o1, Pair<String, RexNode> o2) {
+          return o1.left.compareTo(o2.left);
+        }
+      });
+      return rexBuilder.makeCall(call.getOperator(), Pair.right(newOperands));
```

Comment:
An easier way to achieve this might be to use a SortedMap<String, RexNode>(), then use treeMap.values(). Also, wrap it in a method that can short circuit - you don't need to build a map if there are fewer than 2 operands.